### ISSUE TITLE
Compiled paged attention: masked KV slots change the output

### ISSUE THREAD

**Issue by yannicks1** [torch-spyre/torch-spyre#4517]
### What happened?

In the compiled paged attention kernel, the content of KV slots covered by the additive mask changes the output. Real slots held byte-identical; only the masked trailing slots of a partially-filled block vary (query_len=32, kv_valid=32):

| masked tail | differing elems /131072 | maxabsdiff |
|---|---|---|
| zero vs one  | 75 | 4.88e-4 |
| zero vs real | 143 | 0.0156 |
| zero vs 1e3  | 56170 | 0.0313 |

* **Full block** (`--kv-valid 128`, nothing masked): 0/131072, all four tails.
* **Not saturation**: `nan=0`, `inf=0`, `absmax=15.03` in every arm.
* **Sentinel varied** over `finfo.min`, `finfo.min/2`, `-1000` at a fixed tail: **bit-identical output**. finfo.min/2 is not a fix. Note this control is weaker than it looks: the device's exp clamps every input below ≈−17 to the same value, so exp(-1040), exp(-65504) and exp(-inf) are indistinguishable and the sentinel's magnitude cannot show through whether or not the mask is applied. It does not localise the leak.
* Monotone in masked-slot count: 1/32/64/96 masked → 2/44/84/75 differing (0 → 0).
* Deterministic: three independent processes give byte-identical hashes.

### Root cause
The device's fp16 exp saturates instead of underflowing: every input below ≈−17, -inf included, returns the smallest subnormal 2**-24 (0x0001) where IEEE gives exact zero. Every masked position therefore carries weight 2**-24 and contributes its value through matmul(probs, v) — measured on these shapes, 49152 of 49152 masked lanes were nonzero. Not the gather and not the reduction: a bare matmul is bit-identical when only its unused rows change, on both the output and the reduction axis (0/8192).

### What did you expect to happen?

Bit-identical output when only masked slots change. The kernel already achieves that in two arms — a full block gives 0/131072, and varying the sentinel alone gives 0/131072 — so the partial-block arm is a defect, not fp16 numerics.

### How can we reproduce it?

Standalone — needs only `torch` and `torch_spyre`, no plugin and no model. The paged online-softmax loop is inlined, so the reproducer is the kernel plus its arguments:

    python -u probe.py --num-blocks 1 --kv-valid 32 --query-len 32 --staging-rows 33

Arms: `--kv-valid 128` (control, must be 0/131072), `--kv-valid 127/96/64` (scaling), `--query-len 1` (decode, denominator 4096)

```python
#!/usr/bin/env python3
"""Does the content of MASKED KV slots change the compiled paged-attention output?

Standalone: needs only torch and torch_spyre. The kernel below is the paged
online-softmax attention loop as spyre-inference compiles it, inlined here so the
reproducer carries no plugin dependency.

Real slots are byte-identical across arms; only the masked tail of the last,
partially-filled KV block varies.
"""
import argparse
import hashlib
import sys

import torch
import torch_spyre  # noqa: F401  (registers the "spyre" device; must follow torch)

INT32_ELEMS_PER_STICK = 32


def slot_major_kv_layout(num_slots, num_kv_heads, head_size, dtype):
    """Slot-axis-outermost KV layout. The default tiled layout spreads the slot
    index across two device dims, which makes the indirect store write to the
    wrong rows (torch-spyre#3705)."""
    from torch_spyre._C import SpyreTensorLayout, get_device_dtype, get_elem_in_stick

    eps = get_elem_in_stick(dtype)
    sticks = (head_size + eps - 1) // eps
    return SpyreTensorLayout(
        device_size=[num_slots, num_kv_heads, sticks, eps],
        stride_map=[num_kv_heads * sticks * eps, sticks * eps, eps, 1],
        device_dtype=get_device_dtype(dtype),
    )


def page_attn(query, query_row_index, k_pages, v_pages, page_index_table, mask_tiles,
              scale, num_blocks, padded_query_len, num_heads, num_kv_heads, head_size):
    """Online softmax attention over num_blocks KV pages.

    query:            [num_tokens, num_heads, head_size]
    query_row_index:  int32, first padded_query_len entries are this seq's query rows
    k/v_pages:        [num_pages, block_size, num_kv_heads, head_size]
    page_index_table: [num_blocks, INT32_ELEMS_PER_STICK] int32, page id at column 0
    mask_tiles:       list of num_blocks tensors of [padded_query_len, block_size]
    """
    num_queries_per_kv = num_heads // num_kv_heads
    # A compiled region reads a view from offset 0, ignoring storage_offset
    # (torch-spyre#3770), so the rows are gathered here rather than sliced outside.
    q_rows = query.index_select(0, query_row_index[:padded_query_len])
    q = (q_rows.unsqueeze(0).transpose(1, 2)
         .reshape(num_kv_heads, num_queries_per_kv, padded_query_len, head_size))

    tile_max = tile_sum = tile_output = None
    for i in range(num_blocks):
        # index_select, not k_pages[page_idx]: subscripting lowers to aten.index,
        # which upcasts the int32 index to int64 and fails eager.
        page_idx = page_index_table[i, 0:1]
        k_page = k_pages.index_select(0, page_idx)
        v_page = v_pages.index_select(0, page_idx)
        # Token-major page to head-major for the matmuls; permutes on device.
        k_page_4d = k_page.squeeze(0).permute(1, 0, 2).unsqueeze(1)
        v_page_4d = v_page.squeeze(0).permute(1, 0, 2).unsqueeze(1)

        scores = torch.matmul(q, k_page_4d.transpose(-2, -1)) * scale
        scores = scores + mask_tiles[i]
        scores_max = torch.amax(scores, dim=-1, keepdim=True)
        if i == 0:
            tile_max = scores_max
            tile_probs = torch.exp(scores - tile_max)
            tile_output = torch.matmul(tile_probs, v_page_4d)
            tile_sum = tile_probs.sum(dim=-1, keepdim=True)
        else:
            new_max = torch.maximum(tile_max, scores_max)
            rescale = torch.exp(tile_max - new_max)
            tile_output = tile_output * rescale
            tile_sum = tile_sum * rescale
            tile_probs = torch.exp(scores - new_max)
            tile_output += torch.matmul(tile_probs, v_page_4d)
            tile_sum = tile_sum + tile_probs.sum(dim=-1, keepdim=True)
            tile_max = new_max

    attn = tile_output / tile_sum
    attn = attn.reshape(1, num_heads, padded_query_len, head_size).transpose(1, 2)
    return attn.reshape(padded_query_len, num_heads, head_size)


page_attn_compiled = torch.compile(page_attn, dynamic=False)


def hexhash(t):
    return hashlib.sha256(t.contiguous().view(torch.int16).numpy().tobytes()).hexdigest()[:16]


ap = argparse.ArgumentParser()
ap.add_argument("--num-heads", type=int, default=16)
ap.add_argument("--num-kv-heads", type=int, default=8)
ap.add_argument("--head-size", type=int, default=256)
ap.add_argument("--block-size", type=int, default=128)
ap.add_argument("--num-blocks", type=int, default=1)   # pages the loop walks
ap.add_argument("--num-pages", type=int, default=5)    # pages allocated
ap.add_argument("--query-len", type=int, default=32)
ap.add_argument("--staging-rows", type=int, default=33)
ap.add_argument("--kv-valid", type=int, default=32)    # real slots in the LAST block
ap.add_argument("--scale", type=float, default=1.0)
ap.add_argument("--tails", default="zero,one,real,big100,big1e3")
ap.add_argument("--sentinels", default="min,halfmin,-1000")
args = ap.parse_args()

dtype, dev = torch.float16, torch.device("spyre")
H, KVH, D = args.num_heads, args.num_kv_heads, args.head_size
B, NB, NP = args.block_size, args.num_blocks, args.num_pages
QL, SR, valid = args.query_len, args.staging_rows, args.kv_valid
neg = torch.finfo(dtype).min
# KV pages MUST use slot_major_kv_layout and .to(dev, device_layout=...); a plain
# .to(dev) gives the default tiled layout => "Incompatible host_size and dim_order".
layout = slot_major_kv_layout(NP * B, KVH, D, dtype)
print(f"H={H} KVH={KVH} D={D} block={B} num_blocks={NB} query_len={QL} kv_valid={valid} "
      f"masked_tail={B - valid} sentinel={neg}", flush=True)


def make_kv(tail):
    torch.manual_seed(1234)  # identical REAL content in every arm
    k = torch.randn(NP, B, KVH, D, dtype=dtype).clamp(-1, 1) * 1.7
    v = torch.randn(NP, B, KVH, D, dtype=dtype).clamp(-1, 1) * 15.0
    n = B - valid
    if n > 0:
        if tail == "zero":
            fk = fv = torch.zeros(n, KVH, D, dtype=dtype)
        elif tail == "one":
            fk = fv = torch.ones(n, KVH, D, dtype=dtype)
        elif tail == "real":  # a plausible previous owner's KV
            torch.manual_seed(99)
            fk = torch.randn(n, KVH, D, dtype=dtype).clamp(-1, 1) * 1.7
            fv = torch.randn(n, KVH, D, dtype=dtype).clamp(-1, 1) * 15.0
        else:
            fk = fv = torch.full((n, KVH, D), 100.0 if tail == "big100" else 1000.0, dtype=dtype)
        k[NB, valid:], v[NB, valid:] = fk, fv       # page ids walked are 1..NB
    return k.to(dev, device_layout=layout), v.to(dev, device_layout=layout)


def make_mask(sentinel):
    # A LIST of per-block 2-D tiles: a stacked 5-D tensor sliced on dim 0 fails
    # layout propagation.
    tiles = []
    for b in range(NB):
        tile = torch.zeros(QL, B, dtype=dtype)
        if b == NB - 1 and valid < B:
            tile[:, valid:] = sentinel              # every slot past seq_len
        tiles.append(tile.to(dev))
    return tiles


q = torch.zeros(SR, H, D, dtype=dtype)
torch.manual_seed(7)
q[:QL] = torch.randn(QL, H, D, dtype=dtype) * 0.5
q_staging = q.to(dev)
rows = torch.zeros((QL + INT32_ELEMS_PER_STICK - 1) // INT32_ELEMS_PER_STICK
                   * INT32_ELEMS_PER_STICK, dtype=torch.int32)
rows[:QL] = torch.arange(QL, dtype=torch.int32)
row_table = rows.to(dev)
pit = torch.zeros(NB, INT32_ELEMS_PER_STICK, dtype=torch.int32)
pit[:, 0] = torch.arange(1, NB + 1, dtype=torch.int32)   # page 0 = the null block
page_index_table = pit.to(dev)


def call(kp, vp, tiles):
    out = page_attn_compiled(q_staging, row_table, kp, vp, page_index_table, tiles,
                             args.scale, NB, QL, H, KVH, D)
    return out.to("cpu").clone()[:QL]


def report(kind, arms):
    base = next(iter(arms))
    for name, host in arms.items():
        f = host.float()
        print(f"{kind} {name} hash={hexhash(host)} sum={float(f.double().sum()):.8g} "
              f"absmax={float(f.abs().max()):.6g} nan={int(torch.isnan(f).sum())} "
              f"inf={int(torch.isinf(f).sum())}", flush=True)
    for name, host in arms.items():
        if name == base:
            continue
        ne = int((arms[base] != host).sum())
        print(f"{kind}DIFF {base}-vs-{name} identical={ne == 0} differing_elems={ne}/"
              f"{host.numel()} maxabsdiff="
              f"{float((arms[base].float() - host.float()).abs().max()):.6g}", flush=True)


# (1) vary the MASKED tail, sentinel fixed at finfo.min
tiles = make_mask(neg)
report("MASKLEAK", {t: call(*make_kv(t), tiles) for t in args.tails.split(",") if t})

# (2) vary the SENTINEL, tail fixed -> control that the mask is applied at all
kp, vp = make_kv("real")
sent = {s: (neg if s == "min" else neg / 2 if s == "halfmin" else float(s))
        for s in args.sentinels.split(",") if s}
report("SENTLEAK", {s: call(kp, vp, make_mask(v)) for s, v in sent.items()})

assert not [m for m in sys.modules if m.startswith(("spyre_inference", "vllm"))]
print("PROBE_DONE (no spyre_inference / vllm imported)", flush=True)

```

### Any environmental details we need to know?

-

### Anything else we need to know?

**Impact:** The kernel gathers whole pages, so it reads slots past a sequence's end, and vLLM reuses those blocks for later requests — one forward pass depends on the requests before it. End to end this shows up as a fresh process producing a different answer to the same greedy request. Zeroing the unwritten slots removes it entirely, at a throughput cost (see patch https://github.com/torch-spyre/spyre-inference/pull/873), hence this issue.

### Relevant log output

```Shell
H=16 KVH=8 D=256 block=128 num_blocks=1 query_len=32 kv_valid=32 masked_tail=96 sentinel=-65504.0
MASKLEAK zero   hash=2605ad0da42cfe3d sum=2523.9069 absmax=15.0312 nan=0 inf=0
MASKLEAK one    hash=d16881f794ead246 sum=2523.9084 absmax=15.0312 nan=0 inf=0
MASKLEAK real   hash=80499ed68f0076bb sum=2523.905  absmax=15.0312 nan=0 inf=0
MASKLEAK big100 hash=5958d5a0e975c4f5 sum=2539.1101 absmax=15.0312 nan=0 inf=0
MASKLEAK big1e3 hash=b06283f08cd73707 sum=3024.2466 absmax=15.0312 nan=0 inf=0
MASKLEAKDIFF zero-vs-one    differing_elems=   75/131072 maxabsdiff=0.000488281
MASKLEAKDIFF zero-vs-real   differing_elems=  143/131072 maxabsdiff=0.015625
MASKLEAKDIFF zero-vs-big100 differing_elems= 7825/131072 maxabsdiff=0.03125
MASKLEAKDIFF zero-vs-big1e3 differing_elems=56170/131072 maxabsdiff=0.03125
SENTLEAK min/halfmin/-1000  ALL hash=80499ed68f0076bb
SENTLEAKDIFF min-vs-halfmin differing_elems=0/131072 maxabsdiff=0
SENTLEAKDIFF min-vs--1000   differing_elems=0/131072 maxabsdiff=0

# --kv-valid 128 (full block, nothing masked)
MASKLEAKDIFF zero-vs-one/real/big100/big1e3   all 0/131072

# scaling: masked slots 1 / 32 / 64 / 96
zero-vs-one    2 / 44 / 84 / 75  of 131072
zero-vs-real  16 /108 / 88 /143  of 131072

# --query-len 1
MASKLEAKDIFF zero-vs-one 0/4096   zero-vs-real 1/4096   zero-vs-big1e3 891/4096

# --kv-valid 0 (wholly-masked tile): masked V passes through at weight ~1
MASKLEAK zero   sum=0            absmax=0
MASKLEAK one    sum=131072       absmax=1        (= 131072 x 1.0)
MASKLEAK big100 sum=13107200     absmax=100      (= 131072 x 100)
SENTLEAKDIFF min-vs-halfmin differing_elems=130539/131072 maxabsdiff=24.4062
SENTLEAKDIFF min-vs--1000   differing_elems=125369/131072 maxabsdiff=18.1445

**Comment by dgrove-oss** [comment]
Any chance of a cut-down reproducer that runs without Spyre-inference?

**Comment by yannicks1** [comment]
potential fix: https://github.com/torch-spyre/torch-spyre/pull/4522

**Comment by tdoublep** [comment]
I worked a bit to make this reproducer simpler. I do not believe paged attention is required to illustrate the problem:
```python
import torch
import torch_spyre 

dev, dt = torch.device("spyre"), torch.float16
NEG = torch.finfo(dt).min


def spyre_exp(h):
    """The device's exp of h, as a value and as its raw fp16 bit pattern."""
    d = h.to(dev).exp().to("cpu")
    return float(d[0, 0]), int(d.view(torch.int16)[0, 0]) & 0xFFFF


# (1) The bug: exp() never returns 0, it floors -- at exactly 2**-24, the
#     smallest fp16 subnormal.
for x in (-18.0, -1000.0, NEG):
    h = torch.full((32, 64), x, dtype=dt)
    got, bits = spyre_exp(h)
    print(f"fp16 exp({x:>8g}): cpu={float(h.exp()[0, 0]):<9g} spyre={got:<12g} "
          f"spyre_bits=0x{bits:04x} (expected 0x0000)")

# (2) The consequence: masked slots carry weight 2**-24, so their V reaches the
#     output. Real slots hold 0, so any nonzero output is purely leaked mask.
v = torch.zeros(128, 64, dtype=dt)
v[32:] = 1000.0                                     # only MASKED slots are nonzero
mask = torch.zeros(32, 128, dtype=dt)
mask[:, 32:] = NEG                                  # ... and slots 32.. are masked
leak = (torch.exp(mask.to(dev)) @ v.to(dev)).to("cpu")
print(f"masked V in the output: expected 0, got {float(leak.abs().max()):g}")

# (3) The issue's second symptom: with the whole row masked every weight is the
#     same floor, so the softmax degenerates into an average of the masked V.
p = torch.exp(torch.full((32, 128), NEG, dtype=dt).to(dev))
out = (p @ torch.full((128, 64), 1000.0, dtype=dt).to(dev)) / p.sum(-1, keepdim=True)
print(f"fully-masked softmax:  expected 0, got {float(out.to('cpu').abs().max()):g}")
```
produces:
```
fp16 exp(     -18): cpu=0         spyre=5.96046e-08  spyre_bits=0x0001 (expected 0x0000)
fp16 exp(   -1000): cpu=0         spyre=5.96046e-08  spyre_bits=0x0001 (expected 0x0000)
fp16 exp(  -65504): cpu=0         spyre=5.96046e-08  spyre_bits=0x0001 (expected 0x0000)
masked V in the output: expected 0, got 0.0114899
fully-masked softmax:  expected 0, got 2012
```

**Comment by yannicks1** [comment]
backend fix has been merged by @lupalby . I will close this issue once we have the backend bumped to a version including the fix in spyre-inference. Validated in dev pod that the fix fixes the issue.

**Comment by vswagath1989** [comment]
For record: this is backend PR PR4706

**Comment by yannicks1** [comment]
we bumped the rpms and have this fix in spyre-inference
